### PR TITLE
Fix for #2216, Add Optmizations for ClearText to avoid SSLContext

### PR DESCRIPTION
**Pull request by iNoles**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 1]
```diff
@@ -159,7 +159,16 @@ private OkHttpClient(Builder builder) {
     this.cache = builder.cache;
     this.internalCache = builder.internalCache;
     this.socketFactory = builder.socketFactory;
-    if (builder.sslSocketFactory != null) {
+
+    boolean isTLS = true;
+    for (ConnectionSpec spec : connectionSpecs) {
+      if (connectionSpecs.equals(DEFAULT_CONNECTION_SPECS)) {
+        break;
+      }
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: This check shouldn't be inside the for loop, unless you meant it to be something else.

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 2]
```diff
@@ -159,7 +159,16 @@ private OkHttpClient(Builder builder) {
...
+      isTLS = spec.isTls();
+    }
+
+    if (builder.sslSocketFactory != null || !isTLS) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: It's slightly unfortunate that you always need to loop even when a factory was specified. Trading one inefficiency for another (small) one.

**Comment by iNoles** [comment]
I am trying to find a better way for it. 

Something like
```java
boolean isTLS = true;
if (!connectionSpecs.equals(DEFAULT_CONNECTION_SPECS)) {
    for (ConnectionSpec spec : connectionSpecs) {
        isTLS = spec.isTls();
    }
}
```

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 3]
```diff
@@ -159,7 +159,16 @@ private OkHttpClient(Builder builder) {
...
     this.internalCache = builder.internalCache;
     this.socketFactory = builder.socketFactory;
-    if (builder.sslSocketFactory != null) {
+
+    boolean isTLS = true;
+    for (ConnectionSpec spec : connectionSpecs) {
+      if (connectionSpecs.equals(DEFAULT_CONNECTION_SPECS)) {
+        break;
+      }
+      isTLS = spec.isTls();
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: this needs to be `isTls = isTls || spec.isTls()` because TLS and non-TLS can be mixed.

**Comment by swankjesse** [comment]
One thing I don’t like about this is that it optimizes a case I actively want to discourage. Cleartext connections are bad, and I don’t love the idea of having code to support making the worse thing better.

**Comment by iNoles** [comment]
Do I need to rebase this?

**Comment by swankjesse** [comment]
Nope, looks good. I do request you run http://rebaseandsqua.sh to squish this into a single commit.

**Pull request closed without merging**